versions/development/SPEC.md

```diff
@@ -2327,9 +2327,17 @@ Then the command would be instantiated as:
 /bin/mycmd /path/to/c.txt
 ```
 
-## Prepending a String to an Optional Parameter
+In general, it's invalid to supply an expression of optional type `T?` for a non-optional input or variable of type `T`, as the latter cannot accept `None`. (The idiom `select_first([expr, default])` coerces `expr : T?` to `T` by substituting `default : T` when `expr` is undefined; if `default` is omitted, then the coercion fails at runtime.) Similarly, an `Array[T]` expression does not satisfy an `Array[T]+` input.
 
-Sometimes, optional parameters need a string prefix.  Consider this task:
+These constraints also apply to function arguments, with exceptions for string concatenation, described below, and the equality/inequality operators `==` and `!=`, which can compare two values of types differing only in their quantifiers, considering `None` equal to itself but no other value.
+
+Quantifiers may apply within compound types; for example, if `a : Array[T?]+`, then `a[0] : T?`. An array type can be both non-empty and optional, `Array[T]+?`, such that the value can be `None` or a non-empty array, but not the empty array. (Reliance on this subtle distinction is discouraged, but it arises from other constructs such as a conditional call with an `Array[T]+` output.)
+
+## Interpolating and concatenating optional strings
+
+Interpolations with `~{}` and `${}` accept optional string expressions and substitute the empty string for `None` at runtime.
+
+String concatenation with the `+` operator has special typing properties. When applied to two non-optional operands, the result is a non-optional `String`. However, if either operand has an optional type, then the concatenation has type `String?`, and the runtime result is `None` if either operand is `None`. These unusual semantics facilitate command-line flag interpolation. Consider this task:
```

Comment:
:+1:

Reply:
This was not obvious to me, thank you for pointing it out. For ease of reference, this is the task:

```wdl
task test {
  input {
    String? val
  }
  command {
    python script.py ${"--val=" + val}
  }
}
```

I would much rather not overload the `+` operator to handle optional strings in a special way. Perhaps the die is already cast, because `${"--val=" + val}` is a much loved, common idiom. If at all possible, I would rather use a different symbol than `+`, to avoid overloading.

Reply:
This is technically new, FWIW - right now it's *only* permissible as a special case in string interpolation, and an error elsewhere.

Reply:
I'm fine with the special case although I speculate it's actually a bit trickier to implement, propagating the flag of whether we're inside an interpolation or not when typechecking expressions...

Reply:
This also seems to open the Pandora's box for operators (although that might not be a bad thing), IE should we allow the same logic to be applied to other Types and operators? 
```wdl
Int? one
Int? two

Int? subtract = one - two
Int? add = one + two
Int? multiply = one * two
Int? all = ((one * two) + one) - two
```
To me, it seems like it could be a bad idea to overload the `+` operator and not do the same overloading on all other operators. I think it will likely cause confusion, and I am sure people will inevitably ask for it if in other contexts if we introduce it to string concatenation.